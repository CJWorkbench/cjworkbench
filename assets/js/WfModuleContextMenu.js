// Drop-down menu on Workflows List page, for each listed WF
// triggered by click on three-dot icon next to listed workflow module

import React from 'react'
import {
    UncontrolledDropdown,
    DropdownToggle,
    DropdownMenu,
    DropdownItem,
    Modal,
    ModalHeader,
    ModalBody,
    ModalFooter,
    Form,
    FormGroup,
    Label,
    Input
  } from 'reactstrap'
import PropTypes from 'prop-types'
import CopyToClipboard from 'react-copy-to-clipboard';


export default class WfModuleContextMenu extends React.Component {
  constructor(props) {
    super(props);
    this.deleteOption = this.deleteOption.bind(this);
    this.toggleExportModal = this.toggleExportModal.bind(this);
    this.renderExportModal = this.renderExportModal.bind(this);
    this.onCsvCopy = this.onCsvCopy.bind(this);
    this.onCsvLeave = this.onCsvLeave.bind(this);
    this.onJsonCopy = this.onJsonCopy.bind(this);
    this.onJsonLeave = this.onJsonLeave.bind(this);
    this.state = {
      exportModalOpen: false,
      csvCopied: false,
      jsonCopied: false
    };
  }

  deleteOption() {
    this.props.removeModule();
  }

  toggleExportModal() {
    this.setState({ exportModalOpen: !this.state.exportModalOpen });
  }

  csvUrlString(id) {
    var path = "/public/moduledata/live/" + id + ".csv";
    // allowing an out for testing (there is no window.location.href during test)
    if (window.location.href == 'about:blank') {
      return path;
    } else {
      var url = new URL(path, window.location.href).href;
      return url;
    }
  }

  jsonUrlString(id) {
    var path = "/public/moduledata/live/" + id + ".json";
    // allowing an out for testing (there is no window.location.href during test)
    if (window.location.href == 'about:blank') {
      return path;
    } else {
      var url = new URL(path, window.location.href).href;
      return url;
    }
  }

  onCsvCopy() {
    this.setState({csvCopied: true});
  }

  onCsvLeave() {
    this.setState({csvCopied: false});
  }

  onJsonCopy() {
    this.setState({jsonCopied: true});
  }

  onJsonLeave() {
    this.setState({jsonCopied: false});
  }

  renderCsvCopyLink() {
    var csvString = this.csvUrlString(this.props.id);

    if (this.state.csvCopied) {
      return (
        <div className='info-2 t-orange mt-3' onMouseLeave={this.onCsvLeave}>CSV link copied to clipboard</div>
      );
    } else {
      return (
        <CopyToClipboard text={csvString} onCopy={this.onCsvCopy} className='info-2 t-f-blue mt-3 test-csv-copy'>
          <div>Copy live link</div>
        </CopyToClipboard>
      );
    }
  }

  renderJsonCopyLink() {
    var jsonString = this.jsonUrlString(this.props.id);

    if (this.state.jsonCopied) {
      return (
        <div className='info-2 t-orange mt-3' onMouseLeave={this.onJsonLeave}>JSON link copied to clipboard</div>
      );
    } else {
      return (
        <CopyToClipboard text={jsonString} onCopy={this.onJsonCopy} className='info-2 t-f-blue mt-3 test-json-copy'>
          <div>Copy live link</div>
        </CopyToClipboard>
      );
    }
  }

  renderExportModal() {
    if (!this.state.exportModalOpen) {
      return null;
    }

    var csvString = this.csvUrlString(this.props.id);
    var jsonString = this.jsonUrlString(this.props.id);
    var csvCopyLink = this.renderCsvCopyLink();
    var jsonCopyLink = this.renderJsonCopyLink();

    return (
      <Modal isOpen={this.state.exportModalOpen} toggle={this.toggleExportModal} className={this.props.className}>
        <ModalHeader toggle={this.toggleModal} className='dialog-header modal-header' >
          <div className='t-d-gray title-4'>Export Data</div>
          <div className='icon-close' onClick={this.toggleExportModal}></div>
        </ModalHeader>
        <ModalBody className='dialog-body'>
          <FormGroup>
            <div className='d-flex justify-content-between flex-row'>
              <Label className='t-d-gray info-1'>CSV</Label>
              {csvCopyLink}
            </div>
            <div className='d-flex justify-content-between flex-row mb-3'>
              <Input type='url' className='text-field t-d-gray content-2 test-csv-field' placeholder={csvString} readOnly/>
              <div className='download-icon-box'>
                <a href={csvString} className='icon-download button-icon test-csv-download' download></a>
              </div>
            </div>
            <div className='d-flex justify-content-between flex-row'>
              <Label className='t-d-gray info-1'>JSON</Label>
              {jsonCopyLink}
            </div>
            <div className='d-flex justify-content-between flex-row'>
              <Input type='url' className='text-field t-d-gray content-2 test-json-field' placeholder={jsonString} readOnly/>
              <div className='download-icon-box'>
                <a href={jsonString} className='icon-download button-icon test-json-download' download></a>
              </div>
            </div>
          </FormGroup>
        </ModalBody>
        <ModalFooter className='dialog-footer'>
          <div onClick={this.toggleExportModal} className='button-blue action-button test-done-button'>Done</div>{' '}
        </ModalFooter>
      </Modal>
    );
  }

  render() {
    var exportModal = this.renderExportModal();

    return (
       <UncontrolledDropdown onClick={this.props.stopProp}>
        <DropdownToggle className='context-button'>
          <div className='module-menu-icon icon-more'></div>
        </DropdownToggle>
        <DropdownMenu right className='dropdown-menu'>
          {/* Opens Modal window for downloading files */}
          <DropdownItem key={1} onClick={this.toggleExportModal} className='dropdown-menu-item mb-1 test-export-button'>
            <span className='icon-download button-icon'></span>
            <span className='t-d-gray content-3 ml-3'>Export</span>
            {exportModal}
          </DropdownItem>
          {/* Currently does nothing */}
          <DropdownItem key={2} className='dropdown-menu-item test-update-button'>
            <span className='icon-info button-icon'></span>
            <span className='t-d-gray content-3 ml-3'>Update</span>
          </DropdownItem>
          {/* Will delete the parent WF Module from the list */}
          <DropdownItem key={3} onClick={this.deleteOption} className='dropdown-menu-item test-delete-button'>
            <span className='icon-bin button-icon'></span>
            <span className='t-d-gray content-3 ml-3'>Delete</span>
          </DropdownItem>
        </DropdownMenu>
       </UncontrolledDropdown>
    );
  }
}

WfModuleContextMenu.propTypes = {
  removeModule: PropTypes.func,
  id:           PropTypes.number,
  stopProp:     PropTypes.func
};
