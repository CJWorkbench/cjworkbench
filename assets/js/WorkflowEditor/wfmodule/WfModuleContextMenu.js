// Drop-down menu at upper right of each module in a workflow
import React from 'react'
import PropTypes from 'prop-types'
import { UncontrolledDropdown, DropdownToggle, DropdownMenu, DropdownItem } from '../../components/Dropdown'
import ExportModal from '../../ExportModal'
import { Trans } from '@lingui/macro'

const WfModuleContextMenu = React.memo(function WfModuleContextMenu ({ removeModule, id }) {
  const [isExportModalOpen, setExportModalOpen] = React.useState(false)
  const handleClickOpenExportModal = React.useCallback(() => setExportModalOpen(true))
  const handleCloseExportModal = React.useCallback(() => setExportModalOpen(false))
  const handleClickDelete = React.useCallback(() => removeModule(id))

  return (
    <UncontrolledDropdown>
      <DropdownToggle title='more' className='context-button'>
        <i className='icon-more' />
      </DropdownToggle>
      <DropdownMenu>
        <DropdownItem onClick={handleClickOpenExportModal} className='test-export-button' icon='icon-download'><Trans id='js.WorkflowEditor.wfmodule.wfModuleContextMenu.exportData'>Export data</Trans></DropdownItem>
        <DropdownItem onClick={handleClickDelete} className='test-delete-button' icon='icon-bin'><Trans id='js.WorkflowEditor.wfmodule.wfModuleContextMenu.delete'>Delete</Trans></DropdownItem>
      </DropdownMenu>
      <ExportModal open={isExportModalOpen} wfModuleId={id} toggle={handleCloseExportModal} />
    </UncontrolledDropdown>
  )
})
WfModuleContextMenu.propTypes = {
  removeModule: PropTypes.func,
  id: PropTypes.number
}
export default WfModuleContextMenu
