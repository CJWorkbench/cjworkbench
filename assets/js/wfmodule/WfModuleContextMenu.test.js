import React from 'react'
import WfModuleContextMenu  from './WfModuleContextMenu'
import { mount } from 'enzyme'

describe('WfModuleContextMenu', () => {

  let wrapper;
  let removeModule;
  
  beforeEach(() => {
    removeModule = jest.fn();

    wrapper = mount(
      <WfModuleContextMenu
        removeModule={removeModule} 
        id={415}
        className="menu-test-class"        
      />
    )
  });

  afterEach(() => wrapper.unmount());

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  });

  // only checking the call to removeModule(), not the removal
  it('Renders menu option to Delete with onClick method', () => { 
    // open the context menu
    wrapper.find('button.context-button').simulate('click')

    const deleteButton = wrapper.find('button.test-delete-button');
    expect(deleteButton).toHaveLength(1);
    deleteButton.simulate('click');
    expect(removeModule).toHaveBeenCalled()
  });

  it('should open and close the export modal', () => {
    let modal;

    // open the context menu
    wrapper.find('button.context-button').simulate('click')

    // open the modal window
    const exportButton = wrapper.find('button.test-export-button');
    exportButton.simulate('click');
    expect(wrapper.find('div.modal-dialog')).toHaveLength(1);

    // close it (via callback)
    wrapper.instance().toggleExportModal();
    wrapper.update();
    expect(wrapper.find('div.modal-dialog')).toHaveLength(0);
  });
});
