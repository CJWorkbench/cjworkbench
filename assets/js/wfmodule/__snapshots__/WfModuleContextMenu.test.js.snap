// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WfModuleContextMenu should match snapshot 1`] = `
<WfModuleContextMenu
  className="menu-test-class"
  id={415}
  removeModule={[MockFunction]}
>
  <UncontrolledDropdown>
    <Dropdown
      isOpen={false}
      toggle={[Function]}
    >
      <Manager
        tag={false}
      >
        <div
          className="dropdown"
        >
          <DropdownToggle
            className="context-button"
            title="more"
          >
            <Target>
              <button
                aria-expanded={false}
                className="btn btn-secondary context-button"
                disabled={false}
                onClick={[Function]}
                title="more"
              >
                <i
                  className="context-button--icon icon-more"
                />
              </button>
            </Target>
          </DropdownToggle>
          <DropdownMenu />
          <ExportModal
            onClose={[Function]}
            open={false}
            wfModuleId={415}
          />
        </div>
      </Manager>
    </Dropdown>
  </UncontrolledDropdown>
</WfModuleContextMenu>
`;
